### PR TITLE
kubeadm: do not create etcd datastore if we are in dryrun mode

### LINKED ISSUE #1383: Tests are not passing with unprivileged user
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`): 1.14+

## What happened?

Tests are not passing when run with an unprivileged user that has no rights to mkdir `/var/lib/etcd`.

## What you expected to happen?

Tests should pass.

## How to reproduce it (as minimally and precisely as possible)?

Run tests with unprivileged user that is not able to mkdir `/var/lib/etcd`.

## Anything else we need to know?

The problem was introduced in https://github.com/kubernetes/kubernetes/commit/836f413cf1096c9b020b20319d0767aee4f9b990, because it's trying to `os.MkdirAll(cfg.Etcd.Local.DataDir, 0700)` unconditionally.

### PR DESCRIPTION
**Pull request by ereslibre**
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
This allows for kubeadm tests to pass when run with an unprivileged
user.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/1383

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Review by neolit123 — APPROVED** [review]
thanks @ereslibre 
SGTM
/approve

/priority important-soon
/assign @fabriziopandini @rosti

**Review by aixeshunter — APPROVED** [review]

**Comment by fabriziopandini** [comment]
/approve
I leave @rosti the final lgtm

**Review by rosti — CHANGES_REQUESTED** [review]
@ereslibre thanks for spotting and fixing this!
However we need a couple of more things to be done in both `CreateLocalEtcdStaticPodManifestFile` and `CreateStackedEtcdStaticPodManifestFile` to make them dry run aware.
The calls to `WriteStaticPodToDisk` and `WaitForClusterAvailable` **must not** be called on dry runs, as they'll write the the etcd static pod YAML and wait for it to pop up.
Instead we should call `PrintDryRunFile` to dump the etcd static pod YAML via the dry run log instrumentation for user analysis.

**Comment by ereslibre** [comment]
> The calls to WriteStaticPodToDisk

I was thinking that we could do the same approach as in init:https://github.com/kubernetes/kubernetes/blob/c5da2c559f8451afeb29637c80b0b055a0fb2255/cmd/kubeadm/app/cmd/init.go#L412-L417

> and WaitForClusterAvailable must not be called on dry runs

+1.

WDYT with regards to `WriteStaticPodToDisk`?

**Comment by ereslibre** [comment]
NVM, `join` has no concept of `dryRun` at all so I will go with the straightforward solution for now, it's too cumbersome to include the concept to `join` now.

cc/ @rosti

**Comment by rosti** [comment]
Let's keep it simple for now.
Indeed we don't need dry run support in `CreateStackedEtcdStaticPodManifestFile`, although it would be quite nice to have as we now have the ability to join an additional control plane.

**Comment by ereslibre** [comment]
The current PR is as simple as it gets, didn't touch the stacked logic and we only wrap the `mkdir -p` creation with dry run. This is for fixing unit tests running `kubeadm init --dry-run`, like:

https://github.com/kubernetes/kubernetes/blob/a4e3a5cb52d2e36eff85e10b31e1d4d37afec137/cmd/kubeadm/test/cmd/init_test.go#L34

The current logic of still writing the manifests for `etcd` needs to be out of the dryrun, because we have specific tests for that in `local_test.go` that actually test that the manifest was written in the given directory (in the case of this test a temporary directory):

https://github.com/kubernetes/kubernetes/blob/a4e3a5cb52d2e36eff85e10b31e1d4d37afec137/cmd/kubeadm/app/phases/etcd/local_test.go#L54-L108

So I think we can evolve this later if we want to introduce `--dry-run` for `kubeadm join --experimental-control-plane` and refactor this partial dry running a bit.

**Comment by rosti** [comment]
@ereslibre OK, now I see... There are actually a couple of problems here that need solving. The "dry run" you are actually introducing to `CreateLocalEtcdStaticPodManifestFile` is just a way to avoid creating paths on the root dir during unit tests. The normal dry run is still broken though.

1. Unit tests should override `manifestDir`, `DataDir` and set `dryRun` to `false`.
2. When `dryRun` is `true`, we should simply use `PrintDryRunFile` to dump the etcd static pod YAML and possibly log the `mkdir -p` of the etcd data dir.

The idea behind `kubeadm init --dry-run` is not simply to run some tests. It's to allow sysadmins to check in advance what changes are going to be introduced to their systems.

**Comment by ereslibre** [comment]
@rosti Yes, this should be the best approach. Then I'll fix this PR to do the final desired behavior instead of sneaking a fast test fix and get the proper fix afterwards. Thanks for your feedback.

**Comment by ereslibre** [comment]
> The normal dry run is still broken though.

I don't see how this is the case. `CreateLocalEtcdStaticPodManifestFile` takes a `manifestDir` as a first argument, where it writes the static pod, and this comes from the init data struct (`data.ManifestDir()`).

The dry run on the init case was fine until https://github.com/kubernetes/kubernetes/commit/836f413cf1096c9b020b20319d0767aee4f9b990 got merged. The problem is that this change does not use any information from the init data struct but rather tries to create the etcd data store with safer permissions. What I was trying to achieve with this change was to make tests pass again on top of this change.

**Comment by rosti** [comment]
Ok, I forgot about the dry run dir override in `ManifestDir/init.go`.

> 2. When `dryRun` is `true`, we should simply use `PrintDryRunFile` to dump the etcd static pod YAML and possibly log the `mkdir -p` of the etcd data dir.

Therefore the above part of using `PrintDryRunFile` is an enhancement and bound for another PR. Still it will be nice to drop a log line about the mkdir equivalent we are doing so that we won't surprise anyone.

**Comment by ereslibre** [comment]
> Still it will be nice to drop a log line about the mkdir equivalent we are doing so that we won't surprise anyone.

Thanks, will update the PR!

**Comment by ereslibre** [comment]
@rosti I added the `dryrun` output.

File: cmd/kubeadm/app/phases/etcd/local_test.go
Comment (rosti) [inline cmd/kubeadm/app/phases/etcd/local_test.go]: This needs to be `false` here. The test should not fail.

File: cmd/kubeadm/app/phases/upgrade/staticpods_test.go
Comment (rosti) [inline cmd/kubeadm/app/phases/upgrade/staticpods_test.go]: This needs to be `false` too. In the test the etcd data dir is overridden in the config to a temp dir (see `tmpEtcdDataDir`).

**Review by rosti — COMMENTED** [review]
Thanks @ereslibre ! You are almost there, just a couple of tiny nits.

File: cmd/kubeadm/app/phases/etcd/local_test.go
Comment (ereslibre) [reply cmd/kubeadm/app/phases/etcd/local_test.go]: Thanks, well seen!

**Review by fabriziopandini — COMMENTED** [review]
@ereslibre, thanks for spotting this problem and start working to a solution.

Only one question before reviewing:
Instead of adding a boolean parameter to `CreateLocalEtcdStaticPodManifestFile` and to set it arbitrarly to false in few places, have you considered to move the logic that creates the etcd data store (rif `os.MkdirAll`) outside of this function?

**Comment by ereslibre** [comment]
> Instead of adding a boolean parameter to `CreateLocalEtcdStaticPodManifestFile` and to set it arbitrarly to false in few places, have you considered to move the logic that creates the etcd data store (rif `os.MkdirAll`) outside of this function?

I will explore this option. I went with the less invasive change to fix the unit tests.

**Review by rosti — APPROVED** [review]
Thanks @ereslibre !
I am happy with the way things are right now, however I'll put a hold to allow the latest discussion to go on. Feel free to remove the hold when done.

/lgtm
/hold

**Comment by ereslibre** [comment]
I have this alternative:

https://github.com/kubernetes/kubernetes/compare/master...ereslibre:do-not-create-etcd-datadir-dryrun-alt

@rosti @neolit123 @fabriziopandini thoughts? I wouldn't like to get the `MkdirAll` out of this function TBH, so this is the alternative I came up with. I think given that these functions already accept the `manifestsDir` they could take the `etcdDataDir` as well.

**Comment by neolit123** [comment]
> master...ereslibre:do-not-create-etcd-datadir-dryrun-alt

i think this a good alternative +1 on my side.

**Comment by fabriziopandini** [comment]
@ereslibre I'm still of the idea that getting `mkdir` out of `CreateLocalEtcdStaticPodManifestFile` function is the best solution and the smallest change, but I don't want to block this so I let the final say to @neolit123 / @rosti 

PS The solution in the commit creates a tmp dir even if not necessary and promotes the etcdDir management into `init.go` and `InitData`, that is something that IMO we should avoid

**Comment by neolit123** [comment]
i guess we should move the path creation outside of init.
the etcd phase is the only consumer.

**Comment by ereslibre** [comment]
Okay, updated.

**Review by rosti — APPROVED** [review]
Thanks for the patience and your work @ereslibre !
I like the approach and since we have reached a consensus here, I'll get this going.

/lgtm
/hold cancel

**Pull request merged**
